Tests/Marketplace/search_and_install_packs.py

```diff
@@ -742,82 +561,297 @@ def search_and_install_packs_and_their_dependencies_private(test_pack_path: str,
     return install_packs_private(client, host, pack_ids, test_pack_path)
 
 
-def search_and_install_packs_and_their_dependencies(pack_ids: list,
-                                                    client: demisto_client, hostname: str | None = None,
-                                                    multithreading: bool = True,
-                                                    production_bucket: bool = True):
+def create_graph(
+    all_packs_dependencies: dict,
+) -> DiGraph:
+    """Creates a NetworkX directed graph representing the dependencies between packs.
+
+    Iterates over the provided dictionary containing all packs dependencies and adds
+    directed edges to the graph from each dependent pack to the pack it depends on.
+    Edges are only added for mandatory dependencies.
+
+    Args:
+        all_packs_dependencies (dict): Dictionary containing pack IDs as keys and their
+                                       dependencies as the values.
+
+    Returns:
+        DiGraph: NetworkX directed graph with edges representing dependencies
+                 between packs.
     """
-    Searches for the packs from the specified list, searches their dependencies, and then
-    installs them.
+    graph_dependencies = nx.DiGraph()
+    for pack_id in all_packs_dependencies:
+        pack_dependencies = all_packs_dependencies[pack_id]["dependencies"]
+        for dependence in pack_dependencies:
+            if pack_dependencies[dependence]["mandatory"]:
+                graph_dependencies.add_edge(dependence, pack_id)
+    return graph_dependencies
+
+
+def merge_cycles(graph: DiGraph) -> dict[str, str]:
+    """Merges nodes that form cycles in the directed graph into a single node.
 
     Args:
-        pack_ids (list): A list of the pack ids to search and install.
-        client (demisto_client): The client to connect to.
-        hostname (str): Hostname of instance. Using for logs.
-        multithreading (bool): Whether to use multithreading to install packs in parallel.
-            If multithreading is used, installation requests will be sent in batches of each pack and its dependencies.
-        production_bucket (bool): Whether the installation is in post update mode. Defaults to False.
-    Returns (list, bool):
-        A list of the installed packs' ids, or an empty list if is_nightly == True.
-        A flag that indicates if the operation succeeded or not.
+        graph (DiGraph): The directed graph to merge cycles in.
+
+    Returns:
+        dict[str, str]: A dictionary mapping the original node names that were part of 
+        cycles to the merged node name for that cycle.
+    """
+    logging.debug(
+        f"Found the following cycles in the graph: {list(nx.simple_cycles(graph))}"
+    )
+    map_cycles = {}
+    while list(nx.simple_cycles(graph)):
+        cycle = list(nx.simple_cycles(graph))[0]
```

Comment:
this can be simplified 
```suggestion
    while cycles := list(nx.simple_cycles(graph)):
        cycle = cycles[0]
```